modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala

```diff
@@ -68,44 +68,112 @@ object EnrichmentManager {
-                             )
-      _ = ME.formatContexts(enrichmentsContexts ::: extractResult.validationInfoContexts).foreach(c => enriched.derived_contexts = c)
-      _ <- validateEnriched(
-             enriched,
-             raw,
-             enrichmentsContexts,
-             client,
-             processor,
-             registryLookup,
-             featureFlags.acceptInvalid,
-             invalidCount,
-             atomicFields
-           )
-    } yield enriched
+    atomicFields: AtomicFields,
+    emitIncomplete: Boolean
+  ): IorT[F, BadRow, EnrichedEvent] = {
+    def enrich(enriched: EnrichedEvent): IorT[F, NonEmptyList[NonEmptyList[Failure]], List[SelfDescribingData[Json]]] =
+      for {
+        extractResult <- mapAndValidateInput(
+                           raw,
+                           enriched,
+                           etlTstamp,
+                           processor,
+                           client,
+                           registryLookup
+                         )
+                           .leftMap(NonEmptyList.one)
+                           .possiblyExitingEarly(emitIncomplete)
+        _ = {
+          enriched.contexts = ME.formatContexts(extractResult.contexts).orNull
+          enriched.unstruct_event = ME.formatUnstructEvent(extractResult.unstructEvent).orNull
+        }
+        enrichmentsContexts <- runEnrichments(
+                                 registry,
+                                 raw,
+                                 enriched,
+                                 extractResult.contexts,
+                                 extractResult.unstructEvent,
+                                 featureFlags.legacyEnrichmentOrder
+                               )
+                                 .leftMap(NonEmptyList.one)
+                                 .possiblyExitingEarly(emitIncomplete)
+        validContexts <- validateEnriched(
+                           enriched,
+                           enrichmentsContexts,
+                           client,
+                           registryLookup,
+                           featureFlags.acceptInvalid,
+                           invalidCount,
+                           atomicFields,
+                           emitIncomplete
+                         )
+                           .leftMap(NonEmptyList.one)
+                           .possiblyExitingEarly(emitIncomplete)
+        derivedContexts = validContexts ::: extractResult.validationInfoContexts
+      } yield derivedContexts
+
+    // derived contexts are set lastly because we want to include failure entities
+    // to derived contexts as well and we can get failure entities only in the end
+    // of the enrichment process
+    IorT(
+      for {
+        enrichedEvent <- Sync[F].delay(new EnrichedEvent)
+        enrichmentResult <- enrich(enrichedEvent).value
+        now = Instant.now()
+        _ = setDerivedContexts(enrichedEvent, enrichmentResult, now, processor)
```

Comment:
I don't remember the discussions we had about it early on. 

I think my position on these design aspects was that we should make a best effort for this kind of thing but the implementation difficulty also has a role to play.

If it's not difficult to have it so that the tstamp is different for each, then I would say do it. If it is, then I would say let @stanch worry about whether or not that requirement becomes important enough to spend time on later.

@benjben I would defer to your call on it either way

Reply:
I can't find where but it was agreed that we would use the same timestamp everywhere to simplify the implementation and because we're not sure that having different timestamps is important.

Reply:
Do we have any examples yet of how this timestamp will be used at all?  e.g. any recommended sql query that might use this timestamp for recoveries?

Sorry to keep harping on about this timestamp.  If find it interesting from the point of view of schema design -- and how this timestamp complements the existing `etl_tstamp`.

Reply:
[This is the only note I have recorded about it](https://www.notion.so/keep-in-the-snow/Failure-Entity-Schema-Design-6361207b9801468ca8ac2bbf448c81f0?pvs=4#f8d99bec76b44e3383a009409b11c2ff).

I do vaguely remember the convo we had before about it - if the use case of ordering failures isn't doable then I personally don't feel massively strongly about keeping it or dropping it. 

My only note on the topic is that we should do our best to avoid future breaking changes to this scheme where possible - but I don't think this decision could cause that.

^^ That's my 2c on it anyway! @benjben you'll be much more better with this more recently so I would for sure weigh your input more heavily than mine at this point :)

Reply:
I'm sure that it was agreed on that it was fine to use the same timestamp everywhere, at least for now, but I can't find where on Slack, Github and Notion.

I'd rather avoid to make last minute changes like that, but let me check if it's quick enough to do.

Reply:
Sorry if my last message was unclear - I meant to communicate that I think that was agreed, but my memory of the details of the conversation is hazy, so just offering thoughts on the topic in general.

Apologies if how I phrased it made it sound like I was suggesting we make a change now - that wasn't the intention!

Reply:
No worries at all, I hadn't misinterpreted what you said. But I was still curious about how easy/hard it is to use different timestamps, as we should aim at what the best is. It seems quite straightforward to do, I'm giving it a go.

Reply:
I implemented the change so that each failure gets its own timestamp at the time of creation, could you have a look @istreeter please ?

With the latest implementation it was quite straightforward, which was probably not the case when we decided that it was fine to use the same.